io-uring-test/src/tests/register_sync_cancel.rs

```diff
@@ -0,0 +1,213 @@
+    let completions = wait_get_completions(ring, 2).unwrap();
+    assert_eq!(completions.len(), 2);
+    for completion in completions {
+        assert_eq!(completion.user_data(), USER_DATA_1);
+        assert_eq!(completion.result(), -libc::ECANCELED);
+    }
+
+    // Cancel the fourth and fifth operation by fd.
+    ring.submitter().register_sync_cancel(
+        None,
+        CancelBuilder::new().fd(types::Fd(fd_1.as_raw_fd())).all(),
+    )?;
+    let completions = wait_get_completions(ring, 2).unwrap();
+    assert_eq!(completions.len(), 2);
+    for completion in completions {
+        assert_eq!(completion.user_data(), USER_DATA_2);
+        assert_eq!(completion.result(), -libc::ECANCELED);
+    }
+
+    // Cancel all of the remaining requests.
+    ring.submitter()
+        .register_sync_cancel(None, CancelBuilder::new())?;
+    let completions = wait_get_completions(ring, 1).unwrap();
+    assert_eq!(completions.len(), 3);
+    for completion in completions {
+        assert_eq!(completion.user_data(), USER_DATA_3);
+        assert_eq!(completion.result(), -libc::ECANCELED);
+    }
+
+    Ok(())
+}
+
+pub fn test_register_sync_cancel_any<S: squeue::EntryMarker, C: cqueue::EntryMarker>(
+    ring: &mut IoUring<S, C>,
+    test: &Test,
+) -> io::Result<()> {
+    require!(
+        test;
+        test.probe.is_supported(opcode::SendZc::CODE);
+    );
+    // Test that CancelBuilder::new().all() cancels all requests in the ring.
+    let fd_1 = get_eventfd();
+    const START_USER_DATA: u64 = 47u64;
+    let mut buf = [0u8; 32];
+
+    for i in 0..3 {
+        let entry = opcode::Read::new(types::Fd(fd_1.as_raw_fd()), buf.as_mut_ptr(), 32)
+            .build()
+            .user_data(START_USER_DATA + i);
+        unsafe { ring.submission().push(&entry.into()).unwrap() };
+    }
+    // Submit all 3 operations.
+    assert_eq!(3, ring.submit()?);
+
+    // Cancel all of the requests.
+    ring.submitter()
+        .register_sync_cancel(None, CancelBuilder::new().all())?;
+
+    let completions = wait_get_completions(ring, 5).unwrap();
+    assert_eq!(completions.len(), 3);
+    for completion in completions.iter() {
+        assert_eq!(completion.result(), -libc::ECANCELED);
+    }
+    let mut user_data_entries = completions
+        .into_iter()
+        .map(|c| c.user_data())
+        .collect::<Vec<u64>>();
+    user_data_entries.sort();
+    assert_eq!(
+        user_data_entries,
+        vec![START_USER_DATA, START_USER_DATA + 1, START_USER_DATA + 2,]
+    );
+
+    Ok(())
+}
+
+pub fn test_register_sync_cancel_unsubmitted<S: squeue::EntryMarker, C: cqueue::EntryMarker>(
+    ring: &mut IoUring<S, C>,
+    test: &Test,
+) -> io::Result<()> {
+    require!(
+        test;
+        test.probe.is_supported(opcode::SendZc::CODE);
+    );
+
+    let fd_1 = get_eventfd();
+
+    // Test that we can cancel operations which have not yet been submitted.
```

Comment:
Should this read something like: Try to cancel an operation which has not yet been submitted?